### ISSUE TITLE
Should we define plasma class in a single set of parameters instead of on an array?

### ISSUE THREAD

**Issue by yopology** [PlasmaPy/PlasmaPy#23]
Currently it looks like plasma class is defined on a Cartesian array, that makes it a bit specific. I am thinking maybe we could define a plasma with a single set of parameters e.g. electron/ion number density, temperature, ion charge state, magnetic field, etc., which can be used to calculate Alfven speed, sound speed, Lundquist number, transport coefficients, etc.

And when we have  arrays of non-dimensionalized variables from simulation, which can be in any coordinate system/mesh,  we can use a plasma object to define how to convert non-dimensionalized variables into real units.

Any thoughts on this? This seems a design decision we should make early on.

**Comment by namurphy** [comment]
I've been thinking along the lines of what you're thinking, which is possibly because you're the one who told me about this idea in the first place!  

I have been wondering if it would be best to have multiple classes.  There could be a simple class which exists primarily for calculating plasma parameters (or maybe investigation dispersion relations, etc.) for a set of uniform conditions, without any location information.  We could probably code this generally enough to work with arrays of parameters too (like if we want to calculate Alfven speeds for multiple parcels of plasma at a single time) if the elements of the arrays are treated separately.  That class could perhaps assume fluid/Maxwellian situations where the temperature, etc. can be described with a scalar (or perhaps T_parallel and T_perp), and we could have another class that allows for non-Maxwellian distribution functions but staying with the idea of individual plasma blobs.  In my mind I've been thinking of this like `PlasmaBlob` and `KineticBlob` classes.  

Then we could have separate classes that correspond to fluid/kinetic simulations, including the functionality that you described on converting non-dimensionalized variables into real ones, and also with lots of functionality that can be used to analyze or visualize simulations.  Different simulations would probably require different functions, but could retain the same user interface.  There could also be functionality (in the same or different classes) for fields that are defined analytically.

This seems similar to what you're thinking but am not certain about the best ways to structure the simulation analysis classes.

**Comment by SolarDrew** [comment]
I disagree @yopology , with the disclaimer that I wrote the current Plasma class so I'm obviously biased towards keeping it :)

I think using arrays for parameters gives you more flexibility - you can use an array of a single value for a parameter that's uniform throughout the plasma, whereas you can't use a scalar for a parameter that has a spatial distribution. Also, it's set up in a way that assumes Cartesian at the moment, but in principle I don't see why we couldn't define coordinate information for plasmas in a spherical system, or whatever.

All that said, I'd be happy for the Plasma class itself to be more generic and move a lot of what's there at the moment into a more dedicated Simulation class.

**Comment by namurphy** [comment]
The methods included in a class would likely vary considerably depending on what data (or functions describing physical parameters) are provided in it.  Along the lines of what @yopology is suggesting, there should be a way to do something very simple like:

```Python
blob = PlasmaBlob(B = 0.03*units.T, n = 1.5e17*units.m**-3, T = 1e6*units.K)
blob.print_plasma_parameters()```

where there is no position information whatsoever.  This sort of thing would end up being rather lightweight, and it should be straightforward to enable inputs like `B = [0.2, 0.3, 0.4]*units.T` with other parameters still being perhaps scalars.  It would be great if we could even allow flexible inputs like:

```Python blob = PlasmaBlob(0.03*units.T, 1e6*units.K, 1.5e17*units.m**-3)```

where the class can figure out by units which argument is associated with which field when there is no potential for ambiguity (like, T<sub>i</sub> ≠ T<sub>e</sub>).  

Perhaps the most general `Plasma` class could be used to describe actual plasma configurations, where `PlasmaBlob` could be used solely for things like calculating plasma parameters, etc.  I guess I'm not sure if we should think of the lines above as something different or similar to the `Plasma` class.  @SolarDrew , at one point I think you had mentioned class factories in SunPy, which could be a direction we take this in.  In any case, I have more thinking to do!

**Comment by smithsp** [comment]
From the tokamak point of view, it was disheartening to see the origin classes assuming cartesian geometry.

**Comment by SolarDrew** [comment]
That's fair. I certainly didn't mean to exclude the tokomak community, cartesian is just easier to use for a basic proof of concept. Again, I don't see why we couldn't have some coordinate system information defined as part of the class that would allow the grid you've got to represent some other geometry.

Also, @namurphy, I quite like your idea about flexible function input, but I have no idea how easy that'll be to actually implement robustly.

**Comment by namurphy** [comment]
I was thinking of cartesian geometry as a prototype to make sure we were approaching this from the right angle, but you're right that those of us who primarily use rectangular geometry should start thinking outside the box.

Astropy has a whole subpackage on coordinate transformations and how to go back and forth between equatorial coordinates, galactocentric coordinates, and so on.  That's pretty different than what is needed in plasma physics, but in the long term we'll need to start thinking about representations like Clebsch coordinates, Boozer-Grad coordinates, toroidal flux coordinates, etc. in addition to cylindrical, cartesian, and spherical.  We should definitely aim for as much flexibility as we can for coordinate transformations, probably by including methods in the classes that can do the transformations.  I'm certain that there are existing methods that we can use for things like going between cylindrical and spherical.

[Also if anyone in the future searches this repository for mentions of coordinate systems, a worthwhile reference that we should come back to is _Flux Coordinates and Magnetic Field Structure_ by D'haeseleer, Hitchon, Callen, & Shohet (1991, Springer-Verlag).]

**Comment by lemmatum** [comment]
Would this be a good candidate for inheritance? In other words, make a uniform plasma class (which would have general utility for things like calculating plasma regimes) and then make an array/grid class which inherits the uniform plasma class for each cell (for simulations and stuff). You could probably tweak the inheritance of the latter class to allow for Cartesian, spherical, cylindrical, Klein bottle, Möbius strip, etc. coordinates.

**Comment by smithsp** [comment]
@lemmatum My impression is that having each cell in an array/grid class be its own instance of the uniform plasma class would not be computationally efficient.  Compare numpy array efficiency to looping over each element of the array.  But maybe one could construct the arrays from the cells before doing any math on them.

**Comment by lemmatum** [comment]
Sure, but if the objects are identical, you can make method for converting the parameters contained in an array of objects into a simple numpy array of the parameter. That would give you the options for dealing with objects for certain tests while dealing with arrays for intensive computation. Might even help with coordinate transformations.

**Comment by StanczakDominik** [comment]
I agree with @smithsp, it'd be really slow. As a list of cells. Numpy is the way to go here.

I'll add right here that I have little to no experience with non-uniform grids. There may be things I'm missing in relation to those.

However, suppose we did something like this.

    class PlasmaBlob: # or UniformPlasma
        def __init__(self, T, n):
            self.T = T
            self.n = n
        def physical_quantity(self):
            return self.T * self.n # quick example

This works for floats, straight up numpy arrays and for numpy arrays with units attached to them (thus, quantities) - though for the latter case we could probably pass those into a class inheriting mostly methods from PlasmaBlob, and having access to spatially important methods like gradients. Does that make sense?

Like I said, I have no idea whether this will work with non-uniform grids.

**Comment by smithsp** [comment]
Not sure if you've been down this path, but we have started using `xarray` a lot, and perhaps PlasmaBlob could inherit from `xarray.Dataset`, and then the quantities it stores would be `xarray.DataArrays`, which have dimensions, and the coordinate system could be enforced by the dimensions, with the naming standard enforced by the class.

**Comment by StanczakDominik** [comment]
And that sounds like the solution, basically! If that's a framework that's general enough to deal with non-uniform (and if I'm getting your meaning correctly, also non-cartesian?) grids, I'm all for it. I'll have to start looking into `xarray`.

**Comment by smithsp** [comment]
http://xarray.pydata.org/en/stable/examples/multidimensional-coords.html indicates the coordinates can be separate from the dimensions, supporting non-uniform grids.  However, they have chosen to not tackle unit propagation: http://xarray.pydata.org/en/stable/faq.html?highlight=units#what-is-your-approach-to-metadata

**Comment by smithsp** [comment]
Of course, it appears that this discussion was already had at https://github.com/PlasmaPy/PlasmaPy/issues/59

**Comment by StanczakDominik** [comment]
@smithsp I actually forgot about that until you reminded me, sorry :( But yeah. That's quite the problem with us using xarray, if we're set on units as we seem to be. I'm not quite certain on what the best course of action is here.
